wallets/server.js

```diff
@@ -6,32 +6,55 @@ import * as lnbits from '@/wallets/lnbits/server'
 import * as nwc from '@/wallets/nwc/server'
 import * as phoenixd from '@/wallets/phoenixd/server'
 import * as blink from '@/wallets/blink/server'
+import * as bolt12 from '@/wallets/bolt12/server'
 
 // we import only the metadata of client side wallets
 import * as lnc from '@/wallets/lnc'
 import * as webln from '@/wallets/webln'
 
 import { walletLogger } from '@/api/resolvers/wallet'
 import walletDefs from '@/wallets/server'
-import { parsePaymentRequest } from 'ln-service'
+import { parseInvoice } from '@/lib/boltInvoices'
+import { isBolt12Offer, isBolt12Invoice } from '@/lib/bolt12'
 import { toPositiveBigInt, toPositiveNumber, formatMsats, formatSats, msatsToSats } from '@/lib/format'
 import { PAID_ACTION_TERMINAL_STATES, WALLET_CREATE_INVOICE_TIMEOUT_MS } from '@/lib/constants'
 import { timeoutSignal, withTimeout } from '@/lib/time'
 import { canReceive } from './common'
 import wrapInvoice from './wrap'
 
-export default [lnd, cln, lnAddr, lnbits, nwc, phoenixd, blink, lnc, webln]
+export default [lnd, cln, lnAddr, lnbits, nwc, phoenixd, blink, lnc, webln, bolt12]
 
 const MAX_PENDING_INVOICES_PER_WALLET = 25
 
-export async function createInvoice (userId, { msats, description, descriptionHash, expiry = 360 }, { predecessorId, models }) {
+async function checkInvoice (invoice, { msats }, { lnd, logger }) {
+  const parsedInvoice = await parseInvoice({ lnd, request: invoice })
+  logger.info(`created invoice for ${formatSats(msatsToSats(parsedInvoice.mtokens))}`, {
+    bolt11: invoice
+  })
+  if (BigInt(parsedInvoice.mtokens) !== BigInt(msats)) {
+    if (BigInt(parsedInvoice.mtokens) > BigInt(msats)) {
+      throw new Error('invoice invalid: amount too big')
+    }
+    if (BigInt(parsedInvoice.mtokens) === 0n) {
+      throw new Error('invoice invalid: amount is 0 msats')
+    }
+    if (BigInt(msats) - BigInt(parsedInvoice.mtokens) >= 1000n) {
+      throw new Error('invoice invalid: amount too small')
+    }
+
+    logger.warn('wallet does not support msats')
+  }
+}
+
+export async function createInvoice (userId, { msats, description, descriptionHash, expiry = 360, supportBolt12 = true }, { predecessorId, models, lnd }) {
   // get the wallets in order of priority
   const wallets = await getInvoiceableWallets(userId, { predecessorId, models })
 
   msats = toPositiveNumber(msats)
 
   for (const { def, wallet } of wallets) {
     const logger = walletLogger({ wallet, models })
+    if (def.isBolt12OnlyWallet && !supportBolt12) continue
```

Comment:
Do you still need this? You also check below for `supportBolt12` after the invoice was fetched

Reply:
yes,  this early check is to exclude bolt12-only wallets when WE cannot use them, logging an error for a perfectly fine wallet just because someone is asking an invoice via lnurlp, would be confusing for the user